scripts/create-build-info-package.sh

```diff
@@ -0,0 +1,113 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# This script creates a new package that stores build metadata for packages
+# built by Delphix. It takes one argument: the path where the
+# combined-packages artifacts have been downloaded.
+#
+
+. "${BASH_SOURCE%/*}/common.sh"
+
+if [[ -z "$TOP" ]]; then
+	echo "Must be run inside the git repsitory."
+	exit 1
+fi
+
+if [[ $# -ne 1 ]]; then
+	echo "Must specify the path of the combined-packages artifacts." 1>&2
+	exit 1
+fi
+
+PKG_DIR="$1"
+[[ -d "$PKG_DIR/packages" ]] ||
+	die "$PKG_DIR must be a path to combined-packages artifacts."
+
+#
+# Build info files will be installed into /var/delphix-buildinfo on the
+# Delphix appliance.
+#
+target="$TOP/build-info/var/delphix-buildinfo"
+mkdir -p "$target"
+
+#
+# Copy build metadata for each package
+#
+mkdir "$target/packages"
+cd "$PKG_DIR/packages"
+for pkg in */; do
+	pushd "$pkg" &>/dev/null
+	mkdir "$target/packages/$pkg"
+	for file in GIT_HASH BUILD_INFO PACKAGE_MIRROR_URL_MAIN PACKAGE_MIRROR_URL_SECONDARY metadata.json; do
+		[[ -f "$file" ]] && cp "$file" "$target/packages/$pkg/"
+	done
+	popd &>/dev/null
+done
+
+cp "$PKG_DIR/COMPONENTS" "$target/packages/"
+cp "$PKG_DIR/KERNEL_VERSIONS" "$target/packages/"
+
+#
+# Generate build metadata for appliance-build
+#
+mkdir "$target/appliance-build"
+cd "$TOP"
+git rev-parse HEAD >"$target/appliance-build/GIT_HASH"
+
+function check_env() {
+	#
+	# When the job is ran manually for testing purposes, we do not expect
+	# all environment to be set, so skip the env check.
+	#
+	[[ -n "$JENKINS_URL" ]] || return 0
+
+	local val="${!1}"
+	[[ -n "$val" ]] || die "check_env: $1 must be non-empty"
+	return 0
+}
+
+check_env APPLIANCE_BUILD_GIT_URL
+echo "$APPLIANCE_BUILD_GIT_URL" >"$target/appliance-build/GIT_URL"
+check_env APPLIANCE_BUILD_GIT_BRANCH
+echo "$APPLIANCE_BUILD_GIT_BRANCH" >"$target/appliance-build/GIT_BRANCH"
+check_env DELPHIX_PACKAGE_MIRROR_MAIN
+echo "$DELPHIX_PACKAGE_MIRROR_MAIN" >"$target/appliance-build/DELPHIX_PACKAGE_MIRROR_MAIN"
+check_env DELPHIX_PACKAGE_MIRROR_SECONDARY
+echo "$DELPHIX_PACKAGE_MIRROR_SECONDARY" >"$target/appliance-build/DELPHIX_PACKAGE_MIRROR_SECONDARY"
+check_env AWS_S3_OUTPUT
+echo "$AWS_S3_OUTPUT" >"$target/appliance-build/ARTIFACTS_S3_LOCATION"
+check_env DELPHIX_APPLIANCE_VERSION
+echo "$DELPHIX_APPLIANCE_VERSION" >"$target/appliance-build/DELPHIX_APPLIANCE_VERSION"
+
+#
+# Build the package
+#
+cd "$TOP/build-info"
+#
+# We include some random characters in the version string to avoid collisions
+# with other build-info packages built for different platforms but included in
+# the same upgrade image. The timestamp is insufficient here since the other
+# platforms are built in parallel.
```

Comment:
Can you help me understand this a little better.. as-is, we'll build multiple identical packages for each platform, where the only difference is the package name? why can't we use the same package for all platforms? or, why can't we build a unique/different package for all platforms, but have the name and version be the same for all platforms? or, why can't we use the platform name here, rather than some random bits?

Reply:
> Can you help me understand this a little better.. as-is, we'll build multiple identical packages for each platform, where the only difference is the package name? why can't we use the same package for all platforms?

Having the same package built for all platforms would require us to have it built outside of the script called by appliance-build-stage1, which would require quite some refactoring. Technically we could do something like that for both this package and the ancillary itself repository but I don't think we have a good enough motivation for undertaking this extra work.

>  or, why can't we build a unique/different package for all platforms, but have the name and version be the same for all platforms? or, why can't we use the platform name here, rather than some random bits?

I think this would be quite reasonable actually. I didn't do it because there was no real per-platform differences, but it would be better than adding a random string.

Reply:
> which would require quite some refactoring.

IMO, another reason it'd make more sense to build this package via combine-packages.

Reply:
Since we'll have multiple packages, each with different versions, this means the upgrade image will essentially have this package duplicated for each platform, right?

Reply:
Presumably, a hotfix image will contain the duplicate packages as well? even though this package is irrelevant to any issue being fixed in the hotfix image?

I think even if we refactored to only have a single package (not per-platform), we'd still have to include the package in the hotfix image, even though the package is irrelevant to the hotfix.

Reply:
This package could be useful for the hotfix image as well, since it can be used to validate that the hotfix image indeed contains the packages we want it to contain.

Reply:
Yea, I guess that's true. Can you comment on the size of the package within the upgrade image? Presumably it's on the order of KBs, since it's just a collection of text files? I just want to confirm that it's a trivial size increase to sparse hotfix images.

Reply:
Yeah, the package is tiny, it's 8.8KB.

Reply:
> I think this would be quite reasonable actually. I didn't do it because there was no real per-platform differences, but it would be better than adding a random string.

Actually looking at it more, it would require some additional changes that I'm not sure we want to make. Currently the `ancillaryRepository` gradle task is platform-agnostic, so we'd have to make it per-platform which doesn't seem to make much sense conceptually.

Reply:
> Having the same package built for all platforms would require us to have it built outside of the script called by appliance-build-stage1

Can you clarify why this refactoring would be necessary? just so I understand..

e.g. why can each per-platform build, generate the same package at the same version? if each unique build did this, what would fail?

Reply:
I've had some issues with apt in the past when there were 2 slightly different packages with the same version number, but I don't recall the details. That said, I was curious so I've decided to test it out here just to see what would happen.

I've tried building and image where the timestamp & random digits are removed, and it does seem like the build and the upgrade image generation succeeds (see http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage0/job/pre-push/2422/console). We can see that the aptly repo used for the upgrade image only has a single instance of the delphix-build-info package added.

In practice we'd still want to have a timestamp included in the revision number, so unless we generate a timestamp outside of this script and pass it to all the individual build jobs, we'd likely have multiple build-info packages with slightly different timestamps. I've also compared the produced packages and they are almost the same (the only difference is in the changelog files as they are auto-generated by dch and include a timestamp).

So in short, it seems like we have some other options here, we just need to figure out what we want. I personally don't have much of a preference either way, but if we want to remove the random characters we'd probably want to pass a pre-established timestamp or some other string that is unique to this build but common among all sub-build-jobs.

Reply:
> In practice we'd still want to have a timestamp included in the revision number

Well, I was thinking that maybe we could/should use the `DELPHIX_APPLIANCE_VERSION`.. similar to what we do for `delphix-entire`.. that will automatically include build specific information for builds that are not within the code complete window.. or just the version, e.g. `6.0.9.0`, for builds within code complete.

If we did it this way, we might also need to slightly modify the `execute` script for upgrade, to ensure we always `--reinstall` this package.. similar to what we do for `delphix-entire`..

With that said, I'm not convinced using `DELPHIX_APPLIANCE_VERSION` is better.. just something that I thought about.

Do you have any thoughts about such an approach? I'm fine with what you have, as-is.. just thinking out loud, here..

Reply:
Personally I'd rather just leave the logic as is given how small the extra packages are and given that it should avoid any of those issues.

Reply:
I mean sure, I'm fine with that.. but just to be clear, this is a less than optimal solution (i.e. building and including multiple packages, rather than a single package, in the resultant upgrade and hotfix images).

My approval still stands, so feel free to land.